### ISSUE TITLE
`arc-with-non-send-sync` triggers on macro-generated code

### ISSUE THREAD

**Issue by Nemo157** [rust-lang/rust-clippy#11232]
### Summary

In some external-macro-generated code I have `clippy::arc_with_non_send_sync` warnings getting emitted.

### Reproducer

I tried this code: https://github.com/Nullus157/cbor-diag-rs/blob/7b1726f9545feadc531ab3ed974392ad32ca5aca/tests/proptest.rs#L165-L169 (too big to include locally, and requires a multi-crate example).

I expected to see this happen: no warnings

Instead, this happened:
 
```
error: usage of an `Arc` that is not `Send` or `Sync`
Error:    --> tests/proptest.rs:165:9
    |
165 | /         prop_oneof![
166 | |             arb_array(inner.clone(), 0..10),
167 | |             arb_map(inner.clone(), 0..10),
168 | |             arb_tagged(inner),
169 | |         ]
    | |_________^
    |
    = note: the trait `Send` is not implemented for `impl Strategy<Value = DataItem>`
    = note: the trait `Sync` is not implemented for `impl Strategy<Value = DataItem>`
    = note: required for `Arc<impl Strategy<Value = DataItem>>` to implement `Send` and `Sync`
    = help: consider using an `Rc` instead or wrapping the inner type with a `Mutex`
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#arc_with_non_send_sync
    = note: `-D clippy::arc-with-non-send-sync` implied by `-D warnings`
    = note: this error originates in the macro `$crate::prop_oneof` which comes from the expansion of the macro `prop_oneof` (in Nightly builds, run with -Z macro-backtrace for more info)
```

https://github.com/Nullus157/cbor-diag-rs/actions/runs/5650440647/job/15306784352

### Version

```text
rustc 1.73.0-nightly (8771282d4 2023-07-23)
binary: rustc
commit-hash: 8771282d4e7a5c4569e49d1f878fb3ba90a974d0
commit-date: 2023-07-23
host: x86_64-unknown-linux-gnu
release: 1.73.0-nightly
LLVM version: 16.0.5
```

**Comment by Centri3** [comment]
[Minimized](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=dc6ca442bfede5f20920cae17ef56a91). Just needs a check for whether the expression is from an expansion, then ideally proc macros too.
